### PR TITLE
Improved Stream Management code responsible for generating `<r/>` requests #server-1324

### PR DESCRIPTION
**Pull request by hantu85**

### CODE REVIEW

File: src/main/java/tigase/server/xmppclient/StreamManagementIOProcessor.java [hunk 1]
```diff
@@ -748,13 +763,32 @@ public void ack(int value) {
...
+		/**
+		 * Sets ack request count value
+		 * @param ackRequestCount
+		 */
+		public void setAckRequestCount(int ackRequestCount) {
+			this.ackRequestCount = ackRequestCount;
+		}
+
+		@Deprecated
+		@TigaseDeprecated(removeIn = "9.0.0", since = "8.3.0", note = "Method will not be called any more")
```
Comment (woj-tek) [inline src/main/java/tigase/server/xmppclient/StreamManagementIOProcessor.java:790]: why?

File: src/test/java/tigase/server/xmppclient/StreamManagementIOProcessorTest.java [hunk 2]
```diff
@@ -197,4 +199,37 @@ public void testForceDelayForResumption() {
 				.forEach((OutQueue.Entry e) -> assertFalse(
 						e.getPacketWithStamp().isXMLNSStaticStr(new String[]{"iq", "delay"}, DELAY_XMLNS)));
 	}
+
+	@Test
+	public void testShouldRequestAck()
+			throws NoSuchFieldException, IllegalAccessException, TigaseStringprepException, InterruptedException {
+		StreamManagementIOProcessor processor = new StreamManagementIOProcessor();
+		Field ackRefMinDelayField = StreamManagementIOProcessor.class.getDeclaredField("ack_request_min_delay");
```
Comment (woj-tek) [inline src/test/java/tigase/server/xmppclient/StreamManagementIOProcessorTest.java:207]: Can't we simply make the field protected? It's a config field all in all

File: src/test/java/tigase/server/xmppclient/StreamManagementIOProcessorTest.java [hunk 3]
```diff
@@ -197,4 +199,37 @@ public void testForceDelayForResumption() {
...
+			outQueue.append(p, 2000, 2000);
+			assertEquals(i == 9, processor.shouldRequestAck(null, outQueue));
+		}
+		outQueue.sendingRequest();
+		assertFalse(processor.shouldRequestAck(null, outQueue));
+		for (int i = 0; i < 10; i++) {
+			Packet p = Packet.packetInstance(
+					new Element("message").withElement("body", null, "Test " + UUID.randomUUID().toString()));
+			outQueue.append(p, 2000, 2000);
+			assertFalse(processor.shouldRequestAck(null, outQueue));
```
Comment (woj-tek) [inline src/test/java/tigase/server/xmppclient/StreamManagementIOProcessorTest.java:226]: Nitpick: maybe add a comment here that we are testing second case (i.e. reducing sending `<r/>` for batch processing)? One can figure it out (especially if it's fresh reviewing code and reading intention of the change) but I prefer to have things explicit - makes it easier in the future.

File: src/main/java/tigase/server/xmppclient/StreamManagementIOProcessor.java [hunk 4]
```diff
@@ -527,11 +531,20 @@ public void streamError(XMPPIOService service, StreamError streamErrorName) {
 
 	/**
 	 * Override this method to define a custom behaviour for request ack. The default implementation will request an ack
-	 * if there are more than {@link #ack_request_count} packets waiting, so you probably want to OR your behaviour with
-	 * this.
+	 * if there are more than {@link #ack_request_count} packets waiting since last request for ack and last ack request
+	 * was not sent in last X ms, so you probably want to OR your behaviour with this.
```
[the remark this answers is not in the document]
Comment (hantu85) [reply src/main/java/tigase/server/xmppclient/StreamManagementIOProcessor.java:535]: Yes, that was in the previous version of the comment, so I've kept it.

File: src/main/java/tigase/server/xmppclient/StreamManagementIOProcessor.java [hunk 5]
```diff
@@ -698,7 +713,7 @@ protected void setCounter(int value) {
 		 * server ack request, {@code false} otherwise.
 		 */
 		private boolean shouldCheckTimeout() {
-			return queue.size() > (ack_request_count + 3);
+			return queue.size() > (ackRequestCount + 3);
```
[the remark this answers is not in the document]
Comment (hantu85) [reply src/main/java/tigase/server/xmppclient/StreamManagementIOProcessor.java:716]: Checking timeout is "expensive", so it is better to run it less often (we discussed that when we create this method to overcome stale queues which were long overdue). The idea was to give client some time to respond on `<r/`> request before even trying to check timeouts.

File: src/main/java/tigase/server/xmppclient/StreamManagementIOProcessor.java (hunk 1 above)
Comment (hantu85) [reply src/main/java/tigase/server/xmppclient/StreamManagementIOProcessor.java:790]: As it was setting a variable which was never used. Keeping this is pointess?

File: src/test/java/tigase/server/xmppclient/StreamManagementIOProcessorTest.java (hunk 2 above)
Comment (hantu85) [reply src/test/java/tigase/server/xmppclient/StreamManagementIOProcessorTest.java:207]: we could create getter and setter as well. I've used this method as I do not see a point to expose this field to the other classes.

File: src/main/java/tigase/server/xmppclient/StreamManagementIOProcessor.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (woj-tek) [reply src/main/java/tigase/server/xmppclient/StreamManagementIOProcessor.java:535]: Even though it was in previous version I think we don't have to limit ourselves with "it was like that so let's keep it" and we can improve it still?

[the remark this answers is not in the document]
Comment (hantu85) [reply src/main/java/tigase/server/xmppclient/StreamManagementIOProcessor.java:535]: We merely suggest users use our code with `||` operator with their own condition, is it really bad?

File: src/test/java/tigase/server/xmppclient/StreamManagementIOProcessorTest.java (hunk 3 above)
Comment (woj-tek) [reply src/test/java/tigase/server/xmppclient/StreamManagementIOProcessorTest.java:226]: You write once, you read often - lowering amount of things you have to go through by adding simple comments would be beneficial IMHO, but as I said: nitpick.

**Pull request merged**